sktime/pipeline/_graph_pipe.py

```diff
@@ -0,0 +1,210 @@
+        return self._join_outputs(outputs)
+
+
+class GraphNode(SimpleBaseObject):
+    """GraphStep represents the logic and identity of a single step in GraphPipeline."""
+
+    parameters = {
+        "estimator": BaseEstimator,
+        "name": str,
+        "tags": list,
+        "input_edges": list,
+        "input_indices": list,
+    }
+
+    input_checks = [
+        LengthChecker(
+            to_check=["input_edges", "input_indices"],
+            ignore_none=True,
+        )
+    ]
+
+    def __init__(
+        self,
+        estimator: BaseEstimator,
+        name: str,
+        tags: list = [],
+        input_edges: list = [],
+        input_indices: list = [],
+    ):
+        params = {
+            "estimator": estimator,
+            "name": name,
+            "tags": tags,
+            "input_edges": input_edges,
+            "input_indices": input_indices,
+        }
+        self.run_input_checks(**params)
+        self.set_param(**params)
+
+    def get_estimator(self):
+        """Return the estimator of this GraphNode"""
+        return self.estimator
+
+
+class GraphPipeline(SimpleBaseEstimator):
+    """GraphPipeline facilitates the construction of estimator DAGs.
+
+    This work builds upon work from Benedikt Heidrich (@benHeid)
+    ** Happy to change and add attribution, just putting in the above as
+    a place holder.
+
+    Parameters
+    ----------
+    - steps : list of GraphNode objects, (optional), default = None,
+        contains the GraphNode objects in the order they should be fit.
+
+    Attributes
+    ----------
+
+    - steps : list,
+    - assembled_steps : list,
+    - outputs : GraphOutput,
+
+    Examples
+    --------
+    """
+
+    parameters = {
+        "steps": list,
+        "assembled_steps": list,
+        "estimators": list,
+        "outputs": GraphOutput,
+    }
+
+    @classmethod
+    def assemble(cls, steps: list, outputs=None):
+        estimators = [node.get_estimator() for node in steps]
+        if not outputs:
+            outputs = GraphOutput()
+        est_dict = {}
+        for est in estimators:
+            est_dict[id(est)] = clone(est)
+        assembled_steps = []
+        for node in steps:
+            new_node = GraphNode(
+                est_dict[id(node.estimator)],
+                node.name,
+                node.tags,
+                node.input_edges,
+                node.input_indices,
+            )
+            assembled_steps.append(new_node)
+        params = {
+            "steps": steps,
+            "assembled_steps": assembled_steps,
+            "estimators": estimators,
+            "outputs": outputs,
+        }
+
+    def __init__(
+        self,
+        steps: list = [],
+    ):
+        params = GraphPipeline.assemble(steps)
+        self.set_params(**params)
+
+    def add(
+        self, name: str, estimator: BaseEstimator, input_edges, input_indices, tags
```

Comment:
I am unsure if we need the input_indices besides input_edges. We also should discuss if input_edges is a list or a dict. In the pyWATTS implementation of your API it is currently an API to enable a matching of the predecessors to the parameters of the fit.transform (pyWATTS allows more arguments than x and y).